File: core/model/Account.cpp
//
// Created by liinda on 31.12.18.
//

#include <iostream>
#include "Account.h"
using namespace std;

Account::Account(
  int id, int birth, int joined, Premium *premium, char sex, char status, string phone,
  string email, string *fname, string *sname, string *country, string *city, Like *like, Interest *interest
) :
  id(id), birth(birth), joined(joined), premium(premium), sex(sex), status(status), phone(phone),
  email(email), fname(fname), sname(sname), country(country), city(city), like(like), interest(interest)
  {};

void Account::repr() const {
  std::cout << "<Account: " << endl;
  std::cout << "  " << "id=" << id << endl;
  std::cout << "  " << "birth=" << birth << endl;
  std::cout << "  " << "joined=" << joined << endl;
  std::cout << "  " << "sex=" << sex << endl;
  std::cout << "  " << "status=" << status << endl;
  std::cout << "  " << "phone=" << phone << endl;
  std::cout << "  " << "email=" << email << endl;
  if (fname){std::cout << "  " << "fname=" << *fname << endl;}
  if (sname){std::cout << "  " << "sname=" << *sname << endl;}
  if (country){std::cout << "  " << "country=" << *country << endl;}
  if (city){std::cout << "  " << "city=" << *city << endl;}

  if (premium){cout << "  premium->from=" << premium->from << ", premium->to=" << premium->to << endl;}

  Like *_like = like;
  cout << "  likes:" << endl;
  while (_like != nullptr) {
    cout << "    like->id=" << _like->id << ", like->ts=" << _like->ts << endl;
    _like = _like->next;
  }

  Interest *_interest = interest;
  cout << "  interests:" << endl;
  while (_interest != nullptr) {
    cout << "    interest->name=" << *_interest->name << endl;
    _interest = _interest->next;
  }

  cout << ">" << std::endl;
};

File: src/structures/account.go
package structures

import "fmt"

type Account struct {
	Sex     int8
	Status  int8
	Country int16
	City    int16
	Fname   int16
	Sname   int16
	Domain  int16

	Id     int
	Birth  int
	Joined int

	Email string
	Phone string

	Interests []int16
	Likes     []Likes
	Premium   *Premium
}

func (a1 *Account) Equal(a2 Account) bool {
	if a1.Id == a2.Id &&
		a1.Birth == a2.Birth &&
		a1.Joined == a2.Joined &&
		a1.Country == a2.Country &&
		a1.City == a2.City &&
		a1.Email == a2.Email &&
		a1.Fname == a2.Fname &&
		a1.Sname == a2.Sname &&
		a1.Phone == a2.Phone &&
		a1.Sex == a2.Sex &&
		a1.Status == a2.Status {
		return true
	}
	return false
}

func (acc *Account) Print() {
	fmt.Println("    Id: ", acc.Id)
	fmt.Println("    Sex: ", acc.Sex)
	fmt.Println("    Status: ", acc.Status)
	fmt.Println("    Birth: ", acc.Birth)
	fmt.Println("    Country: ", acc.Country)
	fmt.Println("    City: ", acc.City)
	fmt.Println("    Fname: ", acc.Fname)
	fmt.Println("    Sname: ", acc.Sname)
	fmt.Println("    Email: ", acc.Email)
	fmt.Println("    Domain: ", acc.Domain)
	fmt.Println("    Phone: ", acc.Phone)
	fmt.Println("    Joined: ", acc.Joined)
	fmt.Println("    Interests: ", acc.Interests)
	fmt.Println("    Likes: ", acc.Likes)
	fmt.Println("    Premium: ", acc.Premium)
}

const (
	SexF = "f"
	SexM = "m"
)

func SexIndex(sex string) int8 {
	switch sex {
	case SexF:
		return 0
	case SexM:
		return 1

	default:
		panic("Unexpected sex " + sex)
	}
}

func SexValue(i int8) string {
	sex := [2]string{SexF, SexM}
	return sex[i]
}

const (
	StatFree    = "свободны"
	StatBusy    = "заняты"
	StatComplex = "всё сложно"
)

func StatusIndex(status string) int8 {
	switch status {
	case StatFree:
		return 0
	case StatBusy:
		return 1
	case StatComplex:
		return 2

	default:
		panic("Unexpected status " + status)
	}
}

func StatusValue(i int8) string {
	sex := [3]string{StatFree, StatBusy, StatComplex}
	return sex[i]
}
